ludwig/encoders/text_encoders.py

```diff
@@ -2374,3 +2380,99 @@ def output_shape(self) -> torch.Size:
 
     def get_embedding_layer(self) -> nn.Module:
         return self
+
+
+@DeveloperAPI
+@register_encoder("llm", [TEXT])
+class LLMEncoder(Encoder):
+    def __init__(self, encoder_config: LLMEncoderConfig = None, **kwargs):
+        super().__init__()
+        self.config = encoder_config
+
+        self.model_name = self.config.base_model
+        self.model_config = AutoConfig.from_pretrained(self.config.base_model)
+
+        self.model = load_pretrained_from_config(self.config, model_config=self.model_config)
+        self.curr_device = next(self.model.parameters()).device
+        logger.info("Done.")
+
+        self.context_len = get_context_len(self.model_config)
+
+        # TODO(Arnav): This needs be more flexible to account for RoPE Scaling
+        # When merging input IDs and target IDs for LLM fine-tuning, we want to make sure that the merged tensor is
+        # not longer than the global maximum sequence length. This is provided in the preprocessing config. We never
+        # want to exceed the maximum possible context length so we also check for that.
+        # if self.encoder_config.max_sequence_length:
+        #     max_sequence_length = self.encoder_config.max_sequence_length
+        #     self.max_sequence_length = (
+        #        max_sequence_length if max_sequence_length <= self.context_len else self.context_len
+        #     )
+        # else:
+        print(f"CONTEXT LENGTH: {self.context_len}")
+        self.max_sequence_length = self.context_len
+
+        # Initialize tokenizer
+        self.tokenizer = HFTokenizer(self.config.base_model).tokenizer
+
+        self.attention_masks = None
+
+        clear_data_cache()
+
+    @staticmethod
+    def get_schema_cls() -> Type[BaseEncoderConfig]:
+        return LLMEncoderConfig
+
+    @property
+    def input_shape(self) -> torch.Size:
+        return torch.Size([self.max_sequence_length])
+
+    @property
+    def output_shape(self) -> torch.Size:
+        return torch.Size([self.max_sequence_length, self.model_config.hidden_size])
+
+    def get_embedding_layer(self) -> nn.Module:
+        return self
+
+    def initialize_adapter(self):
+        """If an adapter config is provided, we want to wrap the model with a PEFT model for fine-tuning."""
+        if self.config.adapter:
+            self.model = initialize_adapter(self.model, self.config)
+
+            logger.info("==================================================")
+            logger.info("Trainable Parameter Summary For LLM Encoder Fine-Tuning")
+            logger.info(f"Fine-tuning with adapter: {self.config.adapter.type}")
+            self.model.print_trainable_parameters()
+            logger.info("==================================================")
```

Comment:
nit: Should we potentially consider moving this within initialize_adapter?

Reply:
Sure! I wanted the logs to distinguish between the LLM mode type and the LLM encoder, but we can probably do that more nicely

Reply:
Totally makes sense! Will review the rest soon :)